subprojects/files/src/main/java/org/gradle/api/internal/file/collections/BuildDependenciesOnlyFileCollectionResolveContext.java

```diff
@@ -61,17 +60,10 @@ public FileCollectionResolveContext add(Object element) {
         } else if (element instanceof TaskOutputs) {
             TaskOutputs outputs = (TaskOutputs) element;
             taskContext.add(outputs.getFiles());
-        } else if (element instanceof Closure) {
-            Closure closure = (Closure) element;
-            Object closureResult = closure.call();
-            if (closureResult != null) {
-                add(closureResult);
-            }
         } else if (element instanceof Callable) {
-            Callable callable = (Callable) element;
-            Object callableResult = uncheckedCall(callable);
-            if (callableResult != null) {
-                add(callableResult);
+            Object deferredResult = uncheckedCall((Callable) element);
```

Comment:
❓  shouldn't this use `DeferredUtil`?

Reply:
That would change the behaviour for `Provider<T>` values.

Reply:
Right, and providers here should be handled separately to visit their build dependencies. Out of this PR's scope.

Reply:
Right.